packages/common/src/codec-helpers.ts

```diff
@@ -0,0 +1,137 @@
+import {
+  arrayFromPayloads,
+  errorToFailure,
+  failureToError,
+  fromPayloadsAtIndex,
+  LoadedDataConverter,
+  Payload,
+  PayloadCodec,
+  ProtoFailure,
+  TemporalFailure,
+  toPayloads,
+} from '@temporalio/workflow-common';
+
+export async function decodeFromPayloadsAtIndex<T>(
+  converter: LoadedDataConverter,
+  index: number,
+  payloads?: Payload[] | null
+): Promise<T> {
+  const { payloadConverter, payloadCodec } = converter;
+  return fromPayloadsAtIndex(payloadConverter, index, payloads ? await payloadCodec.decode(payloads) : payloads);
+}
+
+export async function decodeArrayFromPayloads(
+  converter: LoadedDataConverter,
+  content?: Payload[] | null
+): Promise<unknown[]> {
+  const { payloadConverter, payloadCodec } = converter;
+  let decodedPayloads = content;
+  if (content) {
+    decodedPayloads = await payloadCodec.decode(content);
+  }
+  return arrayFromPayloads(payloadConverter, decodedPayloads);
+}
+
+export async function decodeOptionalFailureToOptionalError(
+  converter: LoadedDataConverter,
+  failure: ProtoFailure | undefined | null
+): Promise<TemporalFailure | undefined> {
+  const { payloadConverter, payloadCodec } = converter;
+  return failure ? failureToError(await decodeFailure(payloadCodec, failure), payloadConverter) : undefined;
+}
+
+export async function encodeMapToPayloads<K extends string>(
+  converter: LoadedDataConverter,
+  source: Record<K, any>
+): Promise<Record<K, Payload>> {
+  const { payloadConverter, payloadCodec } = converter;
+  return Object.fromEntries(
+    await Promise.all(
+      Object.entries(source).map(async ([k, v]): Promise<[K, Payload]> => {
+        const [payload] = await payloadCodec.encode([payloadConverter.toPayload(v)]);
+        return [k as K, payload];
+      })
+    )
+  ) as Record<K, Payload>;
+}
+
+export async function encodeToPayload(converter: LoadedDataConverter, value: unknown): Promise<Payload> {
+  const { payloadConverter, payloadCodec } = converter;
+  const [payload] = await payloadCodec.encode([payloadConverter.toPayload(value)]);
+  return payload;
+}
+
+export async function encodeToPayloads(
+  converter: LoadedDataConverter,
+  ...values: unknown[]
+): Promise<Payload[] | undefined> {
+  const { payloadConverter, payloadCodec } = converter;
+  if (values.length === 0) {
+    return undefined;
+  }
+  const payloads = toPayloads(payloadConverter, values);
+  return payloads ? await payloadCodec.encode(payloads) : undefined;
+}
+
+export async function encodeErrorToFailure(dataConverter: LoadedDataConverter, error: unknown): Promise<ProtoFailure> {
+  const { payloadConverter, payloadCodec } = dataConverter;
+  return await encodeFailure(payloadCodec, errorToFailure(error, payloadConverter));
+}
+
+/**
+ * Run `codec.encode()` on the {@link Payload}s in a {@link ProtoFailure}. Mutates `failure`.
+ */
+export async function encodeFailure(codec: PayloadCodec, failure: ProtoFailure): Promise<ProtoFailure> {
```

Comment:
Do you believe the clarity is worth the less flexible use? Eg this:

<img width="906" alt="image" src="https://user-images.githubusercontent.com/251288/154158287-55ca2fe0-201a-459b-9740-b2ac399f2710.png">

Becomes:

<img width="768" alt="image" src="https://user-images.githubusercontent.com/251288/154158390-85ddf4cb-17bd-4a99-b941-4ee9938d0d7b.png">

Reply:
As a general pattern, yes. Because it's internal to the SDK it's acceptable though.
Ideally we'd not mutate and return a new object.

Reply:
Come to think of it, I wouldn't mutate since user code might have reference to the proto failure: https://github.com/temporalio/sdk-typescript/blob/bcc55bd726e71a8f5a134fde6f05706cf34c8bc7/packages/common/src/failure.ts#L55

Reply:
https://github.com/temporalio/sdk-typescript/blob/bcc55bd726e71a8f5a134fde6f05706cf34c8bc7/packages/common/src/failure.ts#L250